### PR TITLE
ROS2 tf2-eigen Reloaded

### PR DESCRIPTION
**Pull request by clalancette**
This is a pull request to make tf2-eigen work under ROS2.  I've tested these changes in CI on Linux, Windows, and OSX; you can see the results below:

- Linux: http://ci.ros2.org/job/ci_linux/2362/
- Windows: http://ci.ros2.org/job/ci_windows/2471/
- OSX: http://ci.ros2.org/job/ci_osx/1816/

If you look closely, you'll see that they are all yellow; that's because there was an unused variable in tf2_eigen-test.cpp, which I've since removed.  It was trivial enough that I didn't feel like waiting for the build farm again for it.  Inline I have more comments on things people pointed out in the previous PR: https://github.com/ros2/geometry2/pull/1

connects to ros2/ros2#342

### CODE REVIEW

File: tf2/include/tf2/convert.h
Comment (clalancette) [inline tf2/include/tf2/convert.h]: As pointed out by Mikael, I've changed this to just be a block comment.

File: tf2/include/tf2/transform_datatypes.h
Comment (clalancette) [inline tf2/include/tf2/transform_datatypes.h:46]: This is new.  I determined that this was already being defined elsewhere (namely time.h), so we can just use that version instead.

File: tf2_eigen/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen/CMakeLists.txt:30]: As pointed out by Mikael in the previous review, I reenabled the test here.

File: tf2_eigen/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen/CMakeLists.txt:34]: Because tf2_eigen now depends on tf2_ros, we have to export the dependencies so downstream users get the right include headers.  This is done here.  This is new.

File: tf2_eigen/include/tf2_eigen/tf2_eigen.h
Comment (clalancette) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.h:30]: As a meta-comment in this file, I changed this to use @dhood's toMsg()/fromMsg() timestamp conversion, which cleans up the file nicely.  I also removed any changes dealing with whitespace or style; this is to take care of @tfoote's concern w.r.t. cherry-picking from the ROS1 geometry2.

File: tf2_eigen/package.xml
Comment (clalancette) [inline tf2_eigen/package.xml]: I'll open a PR against the ROS1 geometry2 to change it to format="2".  Once that is in, we can merge it up with this one.

File: tf2_eigen/test/tf2_eigen-test.cpp
Comment (clalancette) [inline tf2_eigen/test/tf2_eigen-test.cpp]: Both of these fixes can go upstream to ROS1 tf2_eigen.  I'll open a PR for these.

File: tf2_ros/CMakeLists.txt
Comment (clalancette) [inline tf2_ros/CMakeLists.txt:174]: tf2_ros depends on rclcpp, so we have to make sure to export the dependencies for downstream users.

File: tf2_eigen/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen/CMakeLists.txt]: This code is a copy of the code in orocos_kinematics_dynamics that Karsten pointed to, and seems to make finding Eigen on Windows work.  Thanks Karsten.

File: tf2_eigen/test/tf2_eigen-test.cpp
Comment (clalancette) [reply tf2_eigen/test/tf2_eigen-test.cpp]: Actually, it turns out that only the EPS removal can go upstream; the PR is here: https://github.com/ros/geometry2/pull/215 .  The ros.h header is needed upstream to get access to the ros::Time symbol, which is still used to construct Stamped objects in ROS1.

File: tf2_eigen/package.xml
Comment (clalancette) [reply tf2_eigen/package.xml]: Opened up as: https://github.com/ros/geometry2/pull/216

Comment (clalancette) [reply tf2_eigen/package.xml]: Tully merged that into the ros/geometry2 repository, so we can merge up with that to get those fixes in here.  I'll do that once the review here is done.

Comment (mikaelarguedas) [reply tf2_eigen/package.xml]: I think that we should stay in sync with upstream regardless of the status of the opened PRs. Then it's up to the PR opener to rebase their branches accordingly if they happen to conflict.

**Comment by wjwwood** [comment]
This currently needs to have conflicts resolved.

Should it be reviewed? I saw lots of email traffic surrounding tf2 in ROS2, but I wasn't following closely, so I'm uncertain about the state of this.

**Comment by clalancette** [comment]
Yeah, this should be reviewed.  I expect the conflicts to be easy to resolve; I'll do it today.

**Comment by clalancette** [comment]
(I guess using the "Resolve" button in the Github UI does a merge, and not a rebase.  I'll make sure to rebase these commits before pushing)

**Comment by clalancette** [comment]
@ros2/team This is still awaiting review so I can merge.

File: config/FindEigen3.cmake
Comment (wjwwood) [inline config/FindEigen3.cmake]: Where does this file come from? I thought Eigen comes with a CMake config file in more modern releases (like on Xenial). We have a module to find Eigen in our ROS 1 repositories which points this out:

https://github.com/ros/cmake_modules/blob/0.4-devel/cmake/Modules/FindEigen.cmake#L1-L4

File: tf2_eigen/CMakeLists.txt
Comment (wjwwood) [reply tf2_eigen/CMakeLists.txt]: Related to my other comment, is it just on Windows that finding Eigen does not work? Maybe the install of Eigen just needs to be added to the `CMAKE_PREFIX_PATH`.

I'm hesitant to include a module to find Eigen in every package that needs it, but if that's really what's required to move this forward, then we'll have to do it for now.

**Review by wjwwood — APPROVED** [review]
Other than concerns about how we're finding Eigen, lgtm.

File: config/FindEigen3.cmake
Comment (clalancette) [reply config/FindEigen3.cmake]: Yeah, on Linuxen (and I'm assuming OSX), we don't need this.  This is strictly for Windows support.

I also don't love the idea of including this module in every package.  Alternatives here would be to use CMAKE_PREFIX_PATH (I don't know if that will make it work on Windows or not), to include this module one time (in, say, ament), or something else?  @nuclearsandwich , any thoughts here on what the correct way to find Eigen on Windows is?

**Comment by mikaelarguedas** [comment]
Note @clalancette : please rebase this branch and resolve conflicts before "Squash and Merge"ing it. The farm can currently not test this branch because of the conflicts.

**Comment by clalancette** [comment]
I've rebased this, and it should be a fast-forward now.

File: config/FindEigen3.cmake
Comment (Karsten1987) [reply config/FindEigen3.cmake]: are you sure that Eigen is not getting found on the windows machines after the nuget package? This should normally be found.
see here: https://github.com/ros2/robot_state_publisher/blob/ros2/CMakeLists.txt#L13

If we need it, then I propose to open another PR where we revive the cmake_modules package and store the Find*.cmake confiles in a common package.

Comment (clalancette) [reply config/FindEigen3.cmake]: @Karsten1987 You were right, we didn't need that anymore; Windows on CI seems to do the right thing now.  So I've removed FindEigen3.cmake.

That being said, I definitely messed something else up when doing some of my cleanup, so I'm going to run one more set of complete CI jobs on this.  I'll post again once those are complete and happy.

**Comment by clalancette** [comment]
CI jobs with the latest code:

http://ci.ros2.org/job/ci_linux/2426/
http://ci.ros2.org/job/ci_linux-aarch64/57/
http://ci.ros2.org/job/ci_osx/1879/
http://ci.ros2.org/job/ci_windows/2554/

All green (or yellow, but that isn't this PRs fault).  So I think this is ready to merge; I know @wjwwood already approved, but since I made changes I'll wait for another approval before merging.

**Comment by mikaelarguedas** [comment]
I think the new warnings on Windows are caused by this PR: http://ci.ros2.org/job/ci_windows/2554/warnings41Result/new/ and should be adressed before considering merging

**Comment by mikaelarguedas** [comment]
Tip: please use badges as described in the Onboarding document so that it's easy to see which jobs are green without having to click on all of them. Example below:

* Linux  [Build Status](http://ci.ros2.org/job/ci_linux/2426/)
* Linux aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/57/)
* MacOS [Build Status](http://ci.ros2.org/job/ci_osx/1879/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2554/)

**Comment by clalancette** [comment]
Oh, good call, I missed that somehow.  I'll check into those.

File: tf2_ros/include/tf2_ros/buffer_interface.h
Comment (Karsten1987) [inline tf2_ros/include/tf2_ros/buffer_interface.h]: does this function still have to be const return?

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h]: Hm, good point.  I'm guessing we actually *don't* want it to be const return, since it is a translation, and users may want to manipulate the returned object later.  Does that make sense?  If so, I'll remove the const.

Comment (Karsten1987) [reply tf2_ros/include/tf2_ros/buffer_interface.h]: It depends on what you want, but right now in line 61 you return a non-const object, so the const-return has no effect.

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h]: Oh, I see.  All right, well, I'll definitely remove the const then :).  Thanks.

File: tf2_ros/include/tf2_ros/buffer_interface.h
Comment (dhood) [inline tf2_ros/include/tf2_ros/buffer_interface.h:58]: For some context here, when working on https://github.com/ros2/geometry2/pull/2, this signature was chosen for compatibility with `getTimestamp` https://github.com/ros2/geometry2/blob/a0896de571fbc9738b71456f15788cbbed1ca6cd/tf2/include/tf2/convert.h#L62 which @tfoote told me returns const ref to allow for chaining.

In particular, changing this signature will break this line: https://github.com/ros2/geometry2/pull/7/files#diff-bb6bd2d938a7c0ed4397920b3d645fd3R65
There are no tests for tf2_geometry_msgs at the moment and these template instantiations often won't give errors unless you're actually using the code. adding the tests is the next thing on my list, I just haven't gotten to it yet.

It may be that the std::move call can be moved into https://github.com/ros2/geometry2/pull/7/files#diff-bb6bd2d938a7c0ed4397920b3d645fd3R65 to adapt for this change (I am not 100% familiar with the intricacies of this).

The std::move was added to fix that compiler warning, but I admit that it wasn't fully tested on Windows. It might be that the signature can be left as it was with some tweaks for it to work on windows. It could also be that it is the way `fromMsg` is being used in this PR that is causing the errors.

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: @dhood, thanks for the context.  I'm not entirely certain why VC++ is unhappy with that code; I'll take another look and see if I can get a better understanding of what's going on with it.

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: All right, so here is my understanding of the situation (based on [1], and what I think I understand)

A signature like: inline const tf2::TimePoint& fromMsg(...) is returning a reference to an object.  This is only OK if the object is going to stick around after return from the function, i.e. this is a reference to a member value, or something like that.  If the object is *not* going to stick around, then this can lead to a crash, or (worse) later stack corruption.  We get away with it in https://github.com/ros2/geometry2/pull/7/files#diff-bb6bd2d938a7c0ed4397920b3d645fd3R70 because it is returning a reference to a passed-in object, which has a different lifetime.  However, that does *not* hold true for fromMsg(TimePoint) here, because we are actually creating a new object.  Given that, I think we *must* change this to return an object, and not a reference.  @dhood, that also implies we'll have to fix that line of code, and API, that you referenced.  Can someone else comment about whether I'm making sense here?

[1] http://stackoverflow.com/questions/752658/is-the-practice-of-returning-a-c-reference-variable-evil

Comment (dirk-thomas) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: @clalancette :+1:

The function you mentioned with "we get away with it" also needs to be changed:

```
const std::string& getFrameId(const geometry_msgs::Vector3Stamped& t) {return t.header.frame_id;}
```

It is trivial to use this function in a way which leads to accessing invalid memory:

```
const std::string & bad()
{
  geometry_msgs::Vector3Stamped t;
  return getFrameId(t);
}

const std::string & s = bad();
```
Trying to use `s`: :boom:

Comment (dirk-thomas) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: Just for reference: https://github.com/ros2/geometry2/pull/7/files#r111020855

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: So after discussing this at length, it looks like the API change to fromMsg() is what we have to do to be correct.  We'll also have to do further API changes to fix up additional problems that @dirk-thomas pointed out, which I'll open up a separate issue for.

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer_interface.h:58]: https://github.com/ros2/geometry2/issues/10

**Comment by clalancette** [comment]
windows: [Build Status](http://ci.ros2.org/job/ci_windows/2558/)
linux: [Build Status](http://ci.ros2.org/job/ci_linux/2436/)
osx: [Build Status](http://ci.ros2.org/job/ci_osx/1884/)

**Comment by clalancette** [comment]
So all of those builds for this PR are unstable, but it also looks like all of the nightlies went unstable too :(.

**Comment by clalancette** [comment]
windows: [Build Status](http://ci.ros2.org/job/ci_windows/2572/)
linux: [Build Status](http://ci.ros2.org/job/ci_linux/2455/)
osx: [Build Status](http://ci.ros2.org/job/ci_osx/1903/)

**Comment by clalancette** [comment]
So it looks like the builds for this are passing (the one unstable for OSX is the known problem from FastRTPS).  Unless I hear any objections in the next little while, I'm going to merge this up.

**Comment by mikaelarguedas** [comment]
cf https://github.com/ros2/geometry2/issues/8#issuecomment-293987659

**Comment by clalancette** [comment]
Oh, right, I didn't realize there were actually differences at this time.  I'll rebase this as well.

**Comment by clalancette** [comment]
This one is waiting on the merge of indigo-devel into ros2, ros2/geometry2#13 .  Once that is in, I'll merge this one up as well.

**Comment by clalancette** [comment]
One more CI job with the latest changes rebased:

windows: [Build Status](http://ci.ros2.org/job/ci_windows/2632/)
linux: [Build Status](http://ci.ros2.org/job/ci_linux/2521/)
osx: [Build Status](http://ci.ros2.org/job/ci_osx/1986/)

**Comment by clalancette** [comment]
All right, CI passed.  I'm going to merge this up as well.

**Pull request merged**
